taichi/transforms/auto_diff.cpp

```diff
@@ -8,6 +8,81 @@
 #include <typeinfo>
 
 TLANG_NAMESPACE_BEGIN
+class IndependentBlocksJudger : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+
+  void visit(LocalLoadStmt *stmt) override {
+    for (auto &lane : stmt->src.data) {
+      touched_allocas_.insert(lane.var->as<AllocaStmt>());
+    }
+  }
+
+  void visit(LocalStoreStmt *stmt) override {
+    touched_allocas_.insert(stmt->dest->as<AllocaStmt>());
+  }
+
+  void visit(AtomicOpStmt *stmt) override {
+    // We don't need to check the global atomics inside the range for-loops
+    // because
+    // 1. If the range for-loop is innermost, they will be captured by
+    // MakeAdjoint anyway
+    // 2. If the range for-loop is not innermost, they will be processed by
+    // another IndependentBlocksJudger
+    if (is_inside_loop_)
+      return;
+    TI_ASSERT(stmt->dest->is<GlobalPtrStmt>());
+    for (const auto &node : stmt->dest->cast<GlobalPtrStmt>()->snodes.data) {
+      if (node->has_grad()) {
+        qualified_atomics_ = false;
+        break;
+      }
+    }
+  }
+
+  void visit(RangeForStmt *stmt) override {
+    inner_most_loop_ = false;
+    is_inside_loop_ = true;
+    stmt->body->accept(this);
+    is_inside_loop_ = false;
+  }
+
+  static bool run(IRNode *root) {
+    IndependentBlocksJudger Judger;
+    Block *block = root->as<Block>();
+    root->accept(&Judger);
+    std::set<Block *> outside_blocks;
+    // Collect all parent blocks (i.e. outside blocks) of the current block for
+    // local load/store stmt checks
+    for (auto b = block->parent_block(); b; b = b->parent_block()) {
+      if (b)
+        outside_blocks.insert(b);
+    }
+    for (const auto &alloca : Judger.touched_allocas_) {
+      // Test if the alloca belongs to the current block
+      if (outside_blocks.find(alloca->parent) != outside_blocks.end()) {
+        // This block is not an IB since it loads/modifies outside variables
+        Judger.qualified_local_ = false;
```

Comment:
Correct me if I'm wrong, do we still need to save qualified_local_ ? ;) we can early return false here right? ;)

Reply:
Good catch! We can definitely removed it. :-P